### ISSUE TITLE
AMQP over Websockets Certificate Issue

### ISSUE THREAD

**Issue by varwnosg** [Azure/azure-iot-sdks#805]
Hello,

Recenlty, I tried to test the "iothub_client_sample_amqp_websockets" SDK sample. Thus, after building it following the guide in the next link,

https://github.com/Azure/azure-iot-sdks/blob/master/doc/get_started/amqp-c-websockets-run-sample.md

I got the error, "ERR: server's cert didn't look good, X509_V_ERR = 20", which means that OpenSSL trusted list does not include the necessary certificate (Baltimore CyberTrust). Actually, a fresh installation of OpenSSL does not include at all certificates.

Following guides over the WEB on how to add a certificate in OpenSSL under Windows OS with no luck I decided to use the certificate by enabling the "MBED_BUILD_TIMESTAMP" MACRO. So in the file "iothub_client_sample_amqp_websockets.c" I added the path of the "certs.h" and I removed preprocessor's IF as follows,

//#ifdef MBED_BUILD_TIMESTAMP
        // For mbed add the certificate information
        if (IoTHubClient_LL_SetOption(iotHubClientHandle, "TrustedCerts", certificates) != IOTHUB_CLIENT_OK)
        {
            printf("failure to set option \"TrustedCerts\"\r\n");
        }
//#endif // MBED_BUILD_TIMESTAMP

However, the above function call generates an error, thus I decided to use "IoTHubClient_SetOption()" instead of "IoTHubClient_LL_SetOption()". The code now compiles fine but the certificate error is still there. Trying to debug the SDK I realized that the second parameter "TrustedCerts" which is by default in the official SDK release (master branch) returns "IOTHUB_CLIENT_INVALID_ARG" in file "iothub_client_ll_uploadtoblob.c".

CONSOLE OUTPUT:
Starting the IoTHub client sample AMQP over WebSockets...
Info: IoT Hub SDK for C, version 1.0.14
Error: Time:Fri Sep  9 12:24:20 2016 File:X:\XXXXXXX\azure-iot-sdks\c\iothub_cli
ent\src\iothubtransportamqp.c Func:_IoTHubTransportAMQP_SetOption Line:2182 Inva
lid option (TrustedCerts) passed to IoTHubTransportAMQP_SetOption
Error: Time:Fri Sep  9 12:24:20 2016 File:X:\XXXXXXX\azure-iot-sdks\c\iothub_cli
ent\src\iothub_client_ll.c Func:_IoTHubClient_LL_SetOption Line:941 underlying t
ransport failed, returned = IOTHUB_CLIENT_ERROR
Error: Time:Fri Sep  9 12:24:21 2016 File:X:\XXXXXXX\azure-iot-sdks\c\iothub_cli
ent\src\iothub_client.c Func:_IoTHubClient_SetOption Line:691 IoTHubClient_LL_Se
tOption failed
failure to set option "TrustedCerts"
IoTHubClient_SetMessageCallback...successful.
[11953446502:6646] NOTICE: Initial logging level 7
[11953446502:6646] NOTICE: Libwebsockets version: 1.6.0 f1cf5be
[11953446502:6646] NOTICE: IPV6 not compiled in
[11953446502:6646] NOTICE: libev support not compiled in
[11953446502:6646] NOTICE:  ctx mem: 364880 bytes
[11953446502:6866] NOTICE:  canonical_hostname = XXXXXXXXXXXXX
[11953446502:6876] NOTICE:  per-conn mem: 200 + 2126 headers + protocol rx buf
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
Press any key to exit the application.
[11953446502:9887] ERR: server's cert didn't look good, X509_V_ERR = 20: error:0
0000014:lib(0):func(0):SSL lib

SYSTEM ENVIRONMENT:
Windows 7 32-bit

Could you please advice.

**Comment by dcristoloveanu** [comment]
Hi @varwnosg 
I have just seen the same when using SChannel and the WS endpoint. In SChannel terms this is SEC_I_RENEGOTIATE.
We are investigating the issue and will keep this thread updated.

Thanks,
/Dan

**Comment by varwnosg** [comment]
Thanks Dan. I will be waiting!

**Comment by dcristoloveanu** [comment]
@varwnosg 
This bug has been fixed and AMQP over WS functionality should be working now for C.
The fix was checked into the develop branch and will be picked up in our next release.

Thanks!
/Dan

**Comment by varwnosg** [comment]
g

**Comment by varwnosg** [comment]
Hi Dan,

I tried the fix but the sdk was crashing in "xio.c" at,

xio_instance->concrete_xio_handle = xio_instance->io_interface_description->concrete_io_create((void*)xio_create_parameters);

with the error "access violation reading error" because no certificate is passed in the following structure in iothubtransportamqp_websockets.c

XIO_HANDLE getWebSocketsIOTransport(const char* fqdn, int port)
{
	WSIO_CONFIG ws_io_config;
	ws_io_config.host = fqdn;
	ws_io_config.port = port;
	ws_io_config.protocol_name = DEFAULT_WS_PROTOCOL_NAME;
	ws_io_config.relative_path = DEFAULT_WS_RELATIVE_PATH;
	ws_io_config.use_ssl = true;

	return xio_create(wsio_get_interface_description(), &ws_io_config);
}

Thus, I decided to include the library "#include "../../../certs/certs.h" and modify the code as follows,

XIO_HANDLE getWebSocketsIOTransport(const char* fqdn, int port)
{
	WSIO_CONFIG ws_io_config;
	ws_io_config.host = fqdn;
	ws_io_config.port = port;
	ws_io_config.protocol_name = DEFAULT_WS_PROTOCOL_NAME;
	ws_io_config.relative_path = DEFAULT_WS_RELATIVE_PATH;
	ws_io_config.use_ssl = true;
	ws_io_config.trusted_ca = certificates;

	return xio_create(wsio_get_interface_description(), &ws_io_config);
}

After that, everything works fines. Could you please verify the change?

**Comment by dcristoloveanu** [comment]
@varwnosg 
Interesting. I see these changes (removing the extra made it to develop (https://github.com/Azure/azure-iot-sdks/compare/develop).
DId you update the submodules after switching to develop by running git submodule update --init --recursive ?

The changes we pushed in are to pass down the certificates to wsio via the set_option function. I suspect that the azure-c-shared-utility did not get updated for you when you switched to the develop branch.

The change you proposed here would also address the break, but in order to have the TrustedCerts option working across all transports the same we decided to pass the certs down via SetOption.

Let me know if you have other issues after updating to the latest with git submodule --init --update.

Cheers,
/Dan
